lib/modules.nix

```diff
@@ -772,6 +772,23 @@ rec {
       else {};
   };
 
+  /* Evaluates to `value`, type-checking it lazily against `type`, and raising an
+     error if the type doesn't match.
+  */
+  typeChecking = { type, value, pos ? builtins.unsafeGetAttrPos "value" args, loc ? [ "value" ] }@args: let
+    merged = mergeDefinitions loc type [
+      { file = pos.file or unknownModule; inherit value; }
+    ];
+  in merged.mergedValue;
```

Comment:
The names of types describe the output type, and it would be reasonable to expect this to check what the name of the type suggests.
However, as you know, this checks module system definition inputs, and not type outputs. This is not an acceptable implementation; at least not for `lib`.
If you don't want to implement an output checking method on the types, you may choose to move these functions to check-meta, with an explanatory comment.

Reply:
I'm not entirely sure what you mean by type inputs and outputs. Is this about my earlier comment about `mkIf`? I don't really see a solution to that that doesn't involve a deep refactoring of the module system.

Reply:
Yes, `mkIf` is one of the differences between type inputs and outputs.
They're not really types but functions, the way they're part of the module system. We try to write idempotent types, but that's not a guarantee.

> deep refactoring

A simple change would be to add a new method to the types; to consider output checking as a completely new use case for types that's distinct from types' use in the module system.
I don't think the module system's merging and checking should be decoupled, because it's better to check the inputs, for better error messages, and it's better not to check the output and rely on a correct merge implementation, because otherwise, we're checking things twice (or perhaps even quadratically), which is slow for no real benefit.

Reply:
So you're suggesting something like `checkOutput = v: isList v && all elemType.checkOutput v`, for `listOf elemType`? That seems like a lot of code duplication, but I guess we could at least make `checkOutput` default to `check` if not given.

In any case I'll probably just move the functions out of `lib` for this PR.

Reply:
I guess you could call that duplication, yes. They're similar code paths to achieve similar goals, but there's no opportunity for code reuse a the type interface level, and any reuse achieved by a `mkOptionType` helper function would be disproportionally slow, considering how widely this code is going to be called.
So yeah, the implementation is not pretty, but what matters most is how it behaves, and the interface towards users of types and the module system.



> In any case I'll probably just move the functions out of `lib` for this PR.

sgtm



> something like `checkOutput = v: isList v && all elemType.checkOutput v`,

Maybe. This does not allow lazy checking, or nice error messages that can be listed, like the assertions and warnings in NixOS. Can be discussed later.